src/clients/Client.cc

```diff
@@ -455,33 +455,34 @@ sameUrlHosts(const char *url1, const char *url2)
 static void
 purgeEntriesByHeader(HttpRequest *req, const char *reqUrl, Http::Message *rep, Http::HdrType hdr)
 {
-    const char *hdrUrl, *absUrl;
-
-    absUrl = NULL;
-    hdrUrl = rep->header.getStr(hdr);
-    if (hdrUrl == NULL) {
+    const auto hdrUrl = rep->header.getStr(hdr);
+    if (!hdrUrl)
         return;
-    }
 
     /*
      * If the URL is relative, make it absolute so we can find it.
      * If it's absolute, make sure the host parts match to avoid DOS attacks
      * as per RFC 2616 13.10.
      */
+    SBuf absUrlMaker;
+    const char *absUrl;
     if (urlIsRelative(hdrUrl)) {
-        absUrl = urlMakeAbsolute(req, hdrUrl);
-        if (absUrl != NULL) {
-            hdrUrl = absUrl;
+        if (req->method.id() == Http::METHOD_CONNECT)
+            absUrl = hdrUrl; // TODO: merge authority-uri and hdrUrl
+        else if (req->url.getScheme() == AnyP::PROTO_URN)
+            absUrl = req->url.absolute().c_str();
+        else {
+            AnyP::Uri tmpUrl = req->url;
+            tmpUrl.addRelativePath(reqUrl);
```

Comment:
In this particular addRelativePath() use case, we do not really want to _modify_ the existing Uri. We want to create a new one. There are no other use cases yet, and I cannot tell whether most future use cases will need to modify the existing URI when merging (unlike the current use case). Have you considered adjusting addRelativePath()-related API so that the method returns a newly formed AnyP::Uri, and this clone-then-modify code can follow the sketch below?

```C++
absUrlMaker = req->url.mergeWithRelativePath(reqUrl).absolute();
```

I do not insist on any changes here. Just a consideration. You know potential use cases better than I do.

Reply:
The other cases I am aware of have an existing Uri object and are simply add/replace'ing its path.

If there is a performance difference I am happy to change. But otherwise an API that does not rely on callers copying the result back over top of the original Uri object seems better to me (for the present state of things).

Reply:
> The other cases I am aware of have an existing Uri object and are simply add/replace'ing its path.

OK, let's leave as is then.

> If there is a performance difference I am happy to change.

There are too many unknown variables here for me to predict which version would be faster without doing a lot of research. It is your call.

> But otherwise an [edit in place] API seems better to me

If you expect future callers to edit in place and do not have performance concerns about the current "edit in place" API, then we should keep it.

Thank you for evaluating the alternative approach. I consider this change request addressed.